persistence/jdbc/src/main/java/org/infinispan/persistence/jdbc/stringbased/JdbcStringBasedStore.java

```diff
@@ -306,101 +309,122 @@ static class PossibleExpirationNotification {
+                        purgedAmount += doNotify(entriesToBeDeleted, deleteResults, unicastProcessor);
+                     } else {
+                        log.errorf("Expected %d entries to be deleted, but deleteresult contains only %d entries");
                      }
                   }
-
-                  if (list == null || !list.isEmpty()) {
-                     purgedAmount += runBatchAndNotify(list, batchDelete, unicastProcessor);
-                  }
-
-                  if (log.isTraceEnabled()) {
-                     log.tracef("Successfully purged %d rows.", purgedAmount);
-                  }
-                  conn.commit();
-                  unicastProcessor.onComplete();
+               } else {
+                  log.errorf("Expected %d entries to be deleted, but deleteresult is empty");
                }
-            } catch (SQLException e) {
-               log.failedClearingJdbcCacheStore(e);
-               try {
-                  conn.rollback();
-               } catch (SQLException ex) {
-                  log.sqlFailureTxRollback(ex);
+            }
+         } while (!entriesToBeDeleted.isEmpty());
+         if (purgedAmount > 0) {
+            log.infof("Successfully purged %d rows of table %s.", purgedAmount, getTableManager().getDataTableName());
+         }
+         unicastProcessor.onComplete();
+      } catch (SQLException e) {
+         log.failedClearingJdbcCacheStore(e);
+         unicastProcessor.onError(e);
+      }
+   }
+
+   private List<PossibleExpirationNotification> getEntriesToBeDeleted() throws SQLException {
+      int batchSize = this.configuration.maxBatchSize();
+      String sql = this.getTableManager().getSelectOnlyExpiredRowsSql();
+      try (var conn = this.connectionFactory.getConnection()) {
+         conn.setAutoCommit(false);
+         try(PreparedStatement ps = conn.prepareStatement(sql)) {
+            ps.setLong(1, this.timeService.wallClockTime());
+            try(ResultSet rs = ps.executeQuery()) {
+               List<PossibleExpirationNotification> result = new ArrayList<>();
+               while (rs.next() && (batchSize > 0)) {
+                  batchSize--;
+                  String keyStr = rs.getString(2);
+                  long expiryTime = rs.getLong(3);
+                  InputStream inputStream = rs.getBinaryStream(1);
+                  MarshalledValue value = JdbcUtil.unmarshall(inputStream, this.marshaller);
+                  result.add(new PossibleExpirationNotification(expiryTime, keyStr, value));
                }
-               unicastProcessor.onError(e);
-            } finally {
-               JdbcUtil.safeClose(rs);
-               JdbcUtil.safeClose(ps);
-               connectionFactory.releaseConnection(conn);
+               return Collections.unmodifiableList(result);
             }
-         }, "jdbcstringstore-purge");
-         return unicastProcessor;
-      });
+         } finally {
+            conn.commit();
```

Comment:
The commit shouldn't be in a finally block. We don't want to commit if there was some other exception that we don't know about, please move inside the try block as the last line instead. Also this block, shouldn't need to commit anything, this one can just be a rollback.

Reply:
moved